tornettools/stage.py

```diff
@@ -111,28 +111,55 @@ def stage_relays(args):
 
     consensus_paths = get_file_list(args.consensus_path)
     logging.info("Processing {} consensus files from {}...".format(len(consensus_paths), args.consensus_path))
-    relays, min_unix_time, max_unix_time, network_stats = process(num_processes, consensus_paths, parse_consensus, combine_parsed_consensus_results)
+    consensuses = process(num_processes, consensus_paths, parse_consensus, lambda x: x)
+    min_unix_time = min([c['pub_dt'] for c in consensuses])
+    max_unix_time = max([c['pub_dt'] for c in consensuses])
 
     servdesc_paths = get_file_list(args.server_descriptor_path)
     logging.info("Processing {} server descriptor files from {}...".format(len(servdesc_paths), args.server_descriptor_path))
     sdesc_args = [[p, min_unix_time, max_unix_time] for p in servdesc_paths]
-    bandwidths = process(num_processes, sdesc_args, parse_serverdesc, combine_parsed_serverdesc_results)
+    serverdescs = process(num_processes, sdesc_args, parse_serverdesc, lambda x: x)
+
+    families = families_from_serverdescs(serverdescs)
+
+    geo = None
+    if args.geoip_path is not None:
+        geo = GeoIP(args.geoip_path)
+
+    cluster_consensuses(families, geo, consensuses)
+
+    relays = relays_from_consensuses(consensuses)
+    network_stats = network_stats_from_consensuses(consensuses)
+    #timestr = get_time_suffix(min_unix_time, max_unix_time)
+    #logging.info("Found {} total unique relays during {} with a median network size of {} relays".format(len(relays), timestr, network_stats['med_count_total']))
+
+    bandwidths = bandwidths_from_serverdescs(serverdescs)
 
     found_bandwidths = 0
-    for fingerprint in relays:
-        if fingerprint in bandwidths:
-            # overwrite empty bandwidth with parsed bandwidth info
-            relays[fingerprint].bandwidths = bandwidths[fingerprint]
-            found_bandwidths += 1
+    for relay in relays.values():
+        for fingerprint in relay.fingerprints:
+            cluster_bandwidths = []
+            bandwidth = bandwidths.get(fingerprint)
+            if bandwidth is not None:
+                cluster_bandwidths.append({
+                    'bandwidth_capacity': int(bandwidth.max_obs_bw),
+                    'bandwidth_rate': int(median(bandwidth.bw_rates)) if len(bandwidth.bw_rates) > 0 else 0,
+                    'bandwidth_burst': int(median(bandwidth.bw_bursts)) if len(bandwidth.bw_bursts) > 0 else 0,
+                })
+            if len(cluster_bandwidths) > 0:
+                relay.bandwidth_capacity = max(b['bandwidth_capacity'] for b in cluster_bandwidths)
+                relay.bandwidth_rate = max(b['bandwidth_rate'] for b in cluster_bandwidths)
+                relay.bandwidth_burst = max(b['bandwidth_burst'] for b in cluster_bandwidths)
+                found_bandwidths += 1
+            else:
+                relay.bandwidth_capacity = 0
+                relay.bandwidth_rate = 0
+                relay.bandwidth_burst = 0
```

Comment:
This looks wrong to me, because we reset the cluster bandwidths list as we iterate items in the cluster. The "cluster representative relay" will always get the bandwidth info of the last iterated relay in the cluster, rather than computing the bandwidth as the max over all relays in the cluster.

I think we want this instead:

```diff
diff --git a/tornettools/stage.py b/tornettools/stage.py
index e11fe3c..298cdb1 100644
--- a/tornettools/stage.py
+++ b/tornettools/stage.py
@@ -136,9 +136,11 @@ def stage_relays(args):
     bandwidths = bandwidths_from_serverdescs(serverdescs)
 
     found_bandwidths = 0
+    # each 'relay' may actually be many relays clustered into one
     for relay in relays.values():
+        # first we want to collect all bandwidth info we have from everyone in the cluster
+        cluster_bandwidths = []
         for fingerprint in relay.fingerprints:
-            cluster_bandwidths = []
             bandwidth = bandwidths.get(fingerprint)
             if bandwidth is not None:
                 cluster_bandwidths.append({
@@ -146,15 +148,16 @@ def stage_relays(args):
                     'bandwidth_rate': int(median(bandwidth.bw_rates)) if len(bandwidth.bw_rates) > 0 else 0,
                     'bandwidth_burst': int(median(bandwidth.bw_bursts)) if len(bandwidth.bw_bursts) > 0 else 0,
                 })
-            if len(cluster_bandwidths) > 0:
-                relay.bandwidth_capacity = max(b['bandwidth_capacity'] for b in cluster_bandwidths)
-                relay.bandwidth_rate = max(b['bandwidth_rate'] for b in cluster_bandwidths)
-                relay.bandwidth_burst = max(b['bandwidth_burst'] for b in cluster_bandwidths)
-                found_bandwidths += 1
-            else:
-                relay.bandwidth_capacity = 0
-                relay.bandwidth_rate = 0
-                relay.bandwidth_burst = 0
+        # now flatten the bandwidth info for the cluster down into a single bandwidth for the cluster representative
+        if len(cluster_bandwidths) > 0:
+            relay.bandwidth_capacity = max(b['bandwidth_capacity'] for b in cluster_bandwidths)
+            relay.bandwidth_rate = max(b['bandwidth_rate'] for b in cluster_bandwidths)
+            relay.bandwidth_burst = max(b['bandwidth_burst'] for b in cluster_bandwidths)
+            found_bandwidths += 1
+        else:
+            relay.bandwidth_capacity = 0
+            relay.bandwidth_rate = 0
+            relay.bandwidth_burst = 0
 
     logging.info("We found bandwidth information for {} of {} relays".format(found_bandwidths, len(relays)))
     # for (k, v) in sorted(relays.items(), key=lambda kv: kv[1].bandwidths.max_obs_bw):
```